Elevation compat

goncalossilva opened the pull request:
This PR adds elevation compat support in a flexible way via `ElevationDelegate`. It's close to the native implementation, handles the most common use cases like extra padding, margins, etc automatically and it's by far the most versatile approach I've seen so far (in some cases it's not closer to the native elevation on L due to this flexibility).

I strongly advise reading the JavaDoc of each class / method and looking at the samples, but here are some pointers:
1. `ElevationDelegate` should be created by the subclass view that needs elevation (eg. FloatingActionButton or ElevatedToolbar), and some of the methods need proxying (a few are mandatory, but there are others to ease the pain resulted from needing padding in pre-L);
2. `ElevationWrapperDrawable` is a drawables that wraps another drawable (the view's background) and draws the shadow. It has a few formulas based on ~~real shadow math~~ _some general math, some trial and error and a buttload of brute-forcing (using a script plus manually measured data)_ :smile: for calculating the shadow length and alpha in each direction;
3. `ElevationUpdateRunnable` does the heavy lifting of creating the all the edge shaders and corner alpha masks (bitmaps), and is generally run in a background thread after the first run. Initially there were no corner alpha masks -- I would simply draw the generated paths on the Canvas directly. However, this took 10x more time than generating the paths themselves.

![Native on the left, compat on the right](https://dl.dropboxusercontent.com/u/1585962/elevation.png)

goncalossilva committed:
Use 8dp as the default elevation for the FAB

goncalossilva committed:
Improve elevation compat support

Add compat elevation support via ElevationDelegate. All that’s needed
is an elevation value and a corner radius (for non-rectangular views).
ElevationDelegate handles wrapping the background drawable and drawing
the shadow, while simultaneously handling padding and (negative)
margins.

Included is a trivial sample (ElevatedTextView) and a more complex
implementation (FloatingActionButton).

jpmcosta committed:
Bug fix: do not set constant state drawable callback to null

When creating a drawable from a constant state, we cannot set the previous
drawble callback to null, because that drawable is probably still active
and being used by the original drawable.

jpmcosta commented:
I noticed a bug with the shadow on the Emulator. It's a visual glitch, it's better if you see for yourselves (API level 19). Not sure if that is a real problem, though.
I'm not review this PR, since I helped Gonçalo a bit with the integration in the lib (mainly in the FAB and WrapperDrawable) and I also submitted a (minor) bug fix.

goncalossilva commented:
There's a problem in this PR: the `setLayoutParams` override is not smart enough in both `FloatingActionButton` and `ElevationDelegate`.

`setLayoutParams` can be called multiple times during inflation when there are `include` tags. This throws in `FloatingActionButton#setLayoutParams` (because the dimensions are no longer `WRAP_CONTENT` on the second invocation) and sets incorrect margins in `ElevationDelegate#setLayoutParams`. I need to look further into this workflow.

goncalossilva commented:
@jpmcosta I have investigated the glitch you report and in all honesty I think it's an emulator issue (can't reproduce in real devices). However, while investigating the problem, I found an optimization I could do which actually solves the emulator problem as well :grinning: 

I still have to iron out the layout quirks though.

jpmcosta committed:
Avoids crash when setLayoutParams is called multiple times

goncalossilva committed:
Fix elevation corner alpha masks

goncalossilva committed:
Fix corner alpha mask shadow drawing

Each corner slice paint needs a color, otherwise the drawing
results are unpredictable.

goncalossilva committed:
Fix and simplify ElevationDelegate workflow

Background, padding, margins and size are handled in
onAttachedToWindow, instead of overriding a bunch of methods.
This leads to much less black magic going on a much more explicit (and simpler) behaviour.

goncalossilva commented:
Finally, this is in a sane state. There was too much black magic going around about `LayoutParams`, padding, etc, so eventually it occurred to me that simplifying `ElevationDelegate` considerably would fix all the issues and provide a much more explicit API.

Ready for review :+1:

rastislavvasko commented on MaterialLibrary/src/main/java/io/doist/material/drawable/TintDrawable.java:
```diff
@@ -83,6 +83,11 @@ public boolean isStateful() {
 
     @Override
+    public boolean setState(int[] stateSet) {
+        return super.setState(stateSet) | super.superSetState(getState());
```
If I understand this correctly, we're forwarding the wrapped drawable's states to this drawable to trigger `onStateChange` to update the tint. The `superSetState` method feels like a hack, though. Maybe we could change `WrapperDrawable.setState` to something like this (resembling `Drawable.setState`):

```
public boolean setState(int[] stateSet) {
    if (!Arrays.equals(getState(), stateSet)) {
        boolean stateChanged = mWrapperState.mDrawable.setState(stateSet);
        return onStateChange(stateSet) | stateChanged;
    }
    return false;
}
```

And just remove `TintDrawable.setState` and `WrapperDrawable.superSetState`.

There's an unnecessary semicolon on L260. :)

goncalossilva replied:
Pinging @jpmcosta. I honestly don't remember this change, but I recall getting help from João because of the `WrapperDrawable` update. There was indeed a problem that we worked around by doing this.

Will remove the unnecessary semicolon :+1:

jpmcosta replied:
Yeah, let's change it. :+1:

rastislavvasko commented on MaterialLibrary/src/main/java/io/doist/material/elevation/ElevationUpdateRunnable.java:
```diff
@@ -0,0 +1,239 @@
+     * Listen for shadow updates following {@link ElevationUpdateRunnable} runs.
+     */
+    public interface ShadowUpdateListener {
+        void onShadowUpdate(int shadowLeftLength, int shadowTopLength, int shadowRightLenght, int shadowBottomLength,
```
Typo in `shadowRightLenght` (also in `ElevationWrapperDrawable`).

goncalossilva replied:
:+1: Will fix!

rastislavvasko commented on MaterialLibrary/src/main/java/io/doist/material/elevation/ElevationDelegate.java:
```diff
@@ -0,0 +1,360 @@
+    }
+
+    /**
+     * Returns a copy of the {@link View}'s {@link ViewGroup.LayoutParams} excluding the size / margin modifications.
```
Nitpick: it returns the current (and not a copy) of the LayoutParams if `elevationDrawable == null`.

goncalossilva replied:
Nice spotting! These docs were revised almost as many times as the code itself :smirk: Will fix!

goncalossilva replied:
I was going to fix the docs, but instead fixed the method. It should provide a consistent API, ie. always return a copy or never return a copy. Since modifying the real layout params is not a good idea, it now returns a copy at all times.

rastislavvasko commented on MaterialLibrary/src/main/java/io/doist/material/widget/FloatingActionButton.java:
```diff
@@ -1,51 +1,39 @@
 
 public class FloatingActionButton extends ImageButton {
-    private static final int DEFAULT_ELEVATION_DP = 6;
+    public static final String LOG_TAG = FloatingActionButton.class.getSimpleName();
```
This is unused.

goncalossilva replied:
:+1:

rastislavvasko commented on MaterialLibrary/src/main/java/io/doist/material/widget/FloatingActionButton.java:
```diff
@@ -1,51 +1,39 @@
 
-    // Used to create a shadow to fake elevation in pre-L androids.
-    private ElevationManager mElevationManager;
+    private static final int DEFAULT_ELEVATION_DP = 8;
```
I think the default elevation was correct before, at 6dp. Ref: http://www.google.com/design/spec/what-is-material/objects-in-3d-space.html. Maybe it makes more sense to make it into a `R.dimen.fab_default_elevation` anyway?

goncalossilva replied:
Definitely :+1: Maybe AppCompat already has this in their dimensions, I'll check when making this change.

goncalossilva replied:
So… appcompat’s fab elevation is still `8dip` :expressionless: https://cloudup.com/crTXJ3surY6

Will change manually yo `6dip`. When we use a `dimen` for this, I think it should be appcompat's. Hopefully this will be fixed soon.

rastislavvasko commented on MaterialLibrary/src/main/java/io/doist/material/widget/FloatingActionButton.java:
```diff
@@ -91,22 +85,8 @@ private void init(Context context, AttributeSet attrs, int defStyleAttr) {
-                        color = ta.getColorStateList(R.styleable.FloatingActionButton_android_color);
-                    }
-                }
+                color = ta.getColorStateList(R.styleable.FloatingActionButton_android_color);
```
I thought this was going to break providing a raw color (`#abc`), but it works! :smile: :+1:

rastislavvasko commented on MaterialSamples/src/main/java/io/doist/material/sample/ElevatedTextView.java:
```diff
@@ -0,0 +1,137 @@
+
+import io.doist.material.elevation.ElevationDelegate;
+
+public class ElevatedTextView extends TextView implements ElevationDelegate.Host {
```
This needs to be updated, it's broken after the last refactoring. But it ends up being very nice and simple! :)

goncalossilva replied:
Will fix, didn't notice it was broken :+1:

goncalossilva replied:
Woah, I can't believe this passed unnoticed. Fixed :+1:

rastislavvasko commented on MaterialLibrary/src/main/res/values/attrs_elevation_delegate.xml:
```diff
@@ -0,0 +1,15 @@
+    <declare-styleable name="ElevationDelegate">
+        <attr name="elevation" />
+        <attr name="cornerRadius" format="dimension" />
+        <attr name="showShadows">
```
Pretty subjective, but maybe `shownShadows` or `shadowVisibility`? I'm not fond of the verb in an attribute name. :smile:

goncalossilva replied:
Both look good. Do you prefer `shadowVisibility`?

rastislavvasko replied:
I'd go with `shownShadows`.

`shadowVisibility` was the first idea, because it resembles `visibility`, but on second thought that's bad (they're unrelated; former is a flag, second enumone is an enum...). As a plus, `ElevationDelegate` already has the corresponding method named `setShownShadows`. :)

goncalossilva replied:
Agree :+1: Changed to `shownShadows`.

rastislavvasko commented on MaterialLibrary/src/main/java/io/doist/material/elevation/ElevationWrapperDrawable.java:
```diff
@@ -0,0 +1,488 @@
+     * Returns the ambient shadow for the current elevation.
+     */
+    private int getShadowLengthAmbient() {
+        return (int) Math.max(1, mElevation * 3 / 8);
```
Why the 1 as minimum?

goncalossilva replied:
Ambient shadow should always exist (at least according to my tests), as long as the elevation is `> 0` (this method won't be called otherwise). 

A better version would probably be: `Math.ceil(mElevation * 3 / 8)`.

rastislavvasko replied:
Yeah, I was thinking the same (with `ceil`). :+1:

goncalossilva committed:
Remove unnecessary semicolon

goncalossilva committed:
Fix typo

s/shadowRightLenght/shadowRightLength

goncalossilva committed:
Always return a copy of the View’s LayoutParams

goncalossilva committed:
Fix FAB default elevation

According to the latest spec revision, it’s 6dip, not 8dip.

goncalossilva committed:
Fix ElevatedTextView sample

goncalossilva committed:
Rename showShadows attr to shownShadows

goncalossilva committed:
Use Math.ceil() for ambient shadow

goncalossilva committed:
Don’t use imageButton’s style as the default

It has a background, which violates FloatingActionButton’s policy.

goncalossilva committed:
Rewrite logic behind WrapperDrawable#setState

Avoids leaking the logic behind it to subclasses.

goncalossilva committed:
Fix drawable callback not being properly set in ElevationDelegate

View#setBackground(null) must be called *before* the drawable is
wrapped. Otherwise, it will clear the newly set callback and break
invalidation.
